viewing_party/party.py

```diff
@@ -1,23 +1,181 @@
+    ratings_list = []
+    if not user_data["watched"]:
+        return 0.0
+    for movie in user_data["watched"]:
+        ratings_list.append(movie["rating"])
+
+    average_rating = sum(ratings_list) / len(ratings_list)
+    return average_rating
 
+def get_most_watched_genre(user_data):
+    genre_count = {}
+    most_watched_genre = ""
+    if not user_data["watched"]:
+        return None
+
+    for movie in user_data["watched"]:
+        if movie["genre"] not in genre_count:
+            genre_count[movie["genre"]] = 1
+        else:
+            genre_count[movie["genre"]] += 1
+
+    most_watched_genre = max(genre_count, key=genre_count.get)
+    return most_watched_genre
 
-# -----------------------------------------
 # ------------- WAVE 3 --------------------
-# -----------------------------------------
+def get_unique_watched(user_data):
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie["title"])
+
+    friends_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in friends_watched:
+                friends_watched.append(movie["title"])
+
+    user_watched_set = set(user_watched)
+    friends_watched_set = set(friends_watched)
+
+    unique_movies_watched_user_only = user_watched_set - friends_watched_set
+
+    unique_movies_watched_user_only_list = []
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_movies_watched_user_only:
+            unique_movies_watched_user_only_list.append(movie)
+
+    return unique_movies_watched_user_only_list
+
+def get_friends_unique_watched(user_data):
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie["title"])
+
+    friends_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in friends_watched:
+                friends_watched.append(movie["title"])
+
+    user_watched_set = set(user_watched)
+    friends_watched_set = set(friends_watched)
+
+    unique_movies_watched_friends_only = friends_watched_set - user_watched_set
+
+    seen_titles = []
+    unique_movies_watched_friends_only_list = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in unique_movies_watched_friends_only and movie["title"] not in seen_titles:
+                unique_movies_watched_friends_only_list.append(movie)
+                seen_titles.append(movie["title"])
+
+    return unique_movies_watched_friends_only_list
 
-        
-# -----------------------------------------
 # ------------- WAVE 4 --------------------
-# -----------------------------------------
+def get_available_recs(user_data):
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie["title"])
+
+    subscriptions = user_data["subscriptions"]

```

Comment:
We aren't manipulating `user_data["subscriptions"]` in any way, so we don't actually need to save a second reference to a list that already exists, especially since we are only using it once elsewhere!